src/micro_endpoint.c

```diff
@@ -0,0 +1,471 @@
+// Copyright 2023 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <ctype.h>
+
+#include "microp.h"
+#include "mem.h"
+
+static microError *dup_with_prefix(char **dst, const char *prefix, const char *src);
+static void free_endpoint(microEndpoint *ep);
+static void handle_request(natsConnection *nc, natsSubscription *sub, natsMsg *msg, void *closure);
+static microError *new_endpoint(microEndpoint **new_ep, microService *m, const char *prefix, microEndpointConfig *cfg, bool is_internal);
+static void on_drain_complete(void *closure);
+static void release_endpoint(microEndpoint *ep);
+static microEndpoint *retain_endpoint(microEndpoint *ep);
+
+microError *
+micro_new_endpoint(microEndpoint **new_ep, microService *m, const char *prefix, microEndpointConfig *cfg, bool is_internal)
+{
+    microError *err = NULL;
+    microEndpoint *ep = NULL;
+    const char *subj;
+
+    if (!micro_is_valid_name(cfg->Name) || (cfg->Handler == NULL))
+        return micro_ErrorInvalidArg;
+
+    if ((cfg->Subject != NULL) && !micro_is_valid_subject(cfg->Subject))
+        return micro_ErrorInvalidArg;
+
+    subj = nats_IsStringEmpty(cfg->Subject) ? cfg->Name : cfg->Subject;
+
+    ep = NATS_CALLOC(1, sizeof(microEndpoint));
+    if (ep == NULL)
+        return micro_ErrorOutOfMemory;
+    ep->refs = 1;
+    ep->is_monitoring_endpoint = is_internal;
+    ep->m = micro_retain_service(m);
+
+    MICRO_CALL(err, micro_ErrorFromStatus(natsMutex_Create(&ep->endpoint_mu)));
+    MICRO_CALL(err, micro_clone_endpoint_config(&ep->config, cfg));
+    MICRO_CALL(err, dup_with_prefix(&ep->name, prefix, cfg->Name));
+    MICRO_CALL(err, dup_with_prefix(&ep->subject, prefix, subj));
+    if (err != NULL)
+    {
+        free_endpoint(ep);
+        return err;
+    }
+
+    *new_ep = ep;
+    return NULL;
+}
+
+microError *
+micro_start_endpoint(microEndpoint *ep)
+{
+    natsStatus s = NATS_OK;
+    if ((ep->subject == NULL) || (ep->config == NULL) || (ep->config->Handler == NULL))
+        // nothing to do
+        return NULL;
+
+    // reset the stats.
+    memset(&ep->stats, 0, sizeof(ep->stats));
+
```

Comment:
Ok, this is where you need to retain the endpoint (and release if something fails in this function), not in micro_stop_endpoint just before calling drain.

Reply:
It ended up a little more complicated, and perhaps can be further simplified but seems accurate now, and the tests are clean. Still need to add a `Done` callback, but inclined to make it just a sync callback from `microService_Stop` - it would be executed either in the user's thread, or in the OnConnectionClosed callback, in itself async?